File: src/main/java/systems/conduit/main/datastore/backend/RedisBackend.java
package systems.conduit.main.datastore.backend;

import redis.clients.jedis.Jedis;
import systems.conduit.main.datastore.DatastoreHandler;
import systems.conduit.main.datastore.ExpirableBackend;
import systems.conduit.main.datastore.Storable;

import java.util.Map;
import java.util.Optional;

/**
 * @author Innectic
 * @since 1/3/2020
 */
public class RedisBackend implements DatastoreHandler, ExpirableBackend {

    private Jedis client = null;

    @Override
    public void attach(Map<String, Object> meta) {
        client = new Jedis((String) meta.get("host"), (int) meta.get("port"));
        client.auth((String) meta.get("password"));
    }

    @Override
    public void detach() {
        client.disconnect();
        client = null;
    }

    private Optional<Jedis> getClient() {
        return Optional.ofNullable(client);
    }

    @Override
    public void set(String key, String value) {
        getClient().ifPresent(j -> j.set(key, value));
    }

    @Override
    public void set(String key, int value) {
        getClient().ifPresent(j -> j.set(key, String.valueOf(value)));
    }

    @Override
    public void set(String key, float value) {
        getClient().ifPresent(j -> j.set(key, String.valueOf(value)));
    }

    @Override
    public void set(String key, double value) {
        getClient().ifPresent(j -> j.set(key, String.valueOf(value)));
    }

    @Override
    public void set(String key, Storable<?> value) {
        getClient().ifPresent(j -> j.set(key, value.serialize()));
    }

    @Override
    public Optional<Integer> getInt(String key) {
        return getClient().map(j -> {
            try {
                return Integer.valueOf(j.get(key));
            } catch (NumberFormatException ignored) {
                return null;
            }
        });
    }

    @Override
    public Optional<Float> getFloat(String key) {
        return getClient().map(j -> {
            try {
                return Float.valueOf(j.get(key));
            } catch (NumberFormatException ignored) {
                return null;
            }
        });
    }

    @Override
    public Optional<Double> getDouble(String key) {
        return getClient().map(j -> {
            try {
                return Double.valueOf(j.get(key));
            } catch (NumberFormatException ignored) {
                return null;
            }
        });
    }

    @Override
    public Optional<String> getString(String key) {
        return getClient().map(j -> j.get(key));
    }

    @Override
    public <T> Optional<Storable<T>> getCustom(String key) {
        return Optional.empty();  // TODO: Figure out how this is all going to work because it's confusing
    }

    @Override
    public void delete(String key) {

    }

    @Override
    public void set(String key, String value, int duration) {
        getClient().ifPresent(j -> j.setex(key, duration, value));
    }

    @Override
    public void set(String key, int value, int duration) {
        getClient().ifPresent(j -> j.setex(key, duration, String.valueOf(value)));
    }

    @Override
    public void set(String key, float value, int duration) {
        getClient().ifPresent(j -> j.setex(key, duration, String.valueOf(value)));
    }

    @Override
    public void set(String key, double value, int duration) {
        getClient().ifPresent(j -> j.setex(key, duration, String.valueOf(value)));
    }

    @Override
    public void set(String key, Storable<?> value, int duration) {
        getClient().ifPresent(j -> j.setex(key, duration, value.serialize()));
    }
}

File: src/main/kotlin/systems/conduit/core/datastore/backend/RedisBackend.kt
package systems.conduit.core.datastore.backend

import redis.clients.jedis.Jedis
import systems.conduit.core.datastore.DatastoreHandler
import systems.conduit.core.datastore.ExpirableBackend
import systems.conduit.core.datastore.Storable
import java.util.*

/**
 * An expirable datastore that holds information in a Redis database.
 *
 * @author Innectic
 * @since 1/3/2020
 */
class RedisBackend: DatastoreHandler, ExpirableBackend {

    var client: Jedis? = null
        private set

    override fun attach(meta: Map<String, Any>) {
        client = Jedis(meta["host"] as String, meta["port"] as Int)
        client!!.auth(meta["password"] as String)
    }

    override fun detach() {
        client!!.disconnect()
        client = null
    }

    override fun set(key: String, value: String) {
        client?.set(key, value)
    }

    override fun set(key: String, value: Int) {
        client?.set(key, value.toString())
    }

    override fun set(key: String, value: Float) {
        client?.set(key, value.toString())
    }

    override fun set(key: String, value: Double) {
        client?.set(key, value.toString())
    }

    override fun set(key: String, value: Storable<*>) {
        client?.set(key, value.toString())
    }

    override fun set(key: String, value: String, duration: Int) {
        client?.setex(key, duration, value)
    }

    override fun set(key: String, value: Int, duration: Int) {
        client?.setex(key, duration, value.toString())
    }

    override fun set(key: String, value: Float, duration: Int) {
        client?.setex(key, duration, value.toString())
    }

    override fun set(key: String, value: Double, duration: Int) {
        client?.setex(key, duration, value.toString())
    }

    override fun set(key: String, value: Storable<*>, duration: Int) {
        client?.setex(key, duration, value.serialize())
    }

    override fun getInt(key: String): Int? {
        return try {
            Integer.valueOf(client?.get(key))
        } catch (ignored: NumberFormatException) {
            null
        }
    }

    override fun getFloat(key: String): Float? {
        return try {
            java.lang.Float.valueOf(client?.get(key))
        } catch (ignored: NumberFormatException) {
            null
        }
    }

    override fun getDouble(key: String): Double? {
        return try {
            java.lang.Double.valueOf(client?.get(key))
        } catch (ignored: NumberFormatException) {
            null
        }
    }

    override fun getString(key: String): String? {
        return client?.get(key)
    }

    override fun <T> getCustom(key: String): Storable<T>? {
        return null // TODO: Figure out how this is all going to work because it's confusing
    }

    override fun delete(key: String) {
        client?.del(key)
    }
}
